src/Symfony/Component/Security/Csrf/DoubleSubmitCsrfTokenManager.php

```diff
@@ -0,0 +1,201 @@
+    {
+        return new CsrfToken($tokenId, $this->cookieName);
+    }
+
+    public function refreshToken(string $tokenId): CsrfToken
+    {
+        return new CsrfToken($tokenId, $this->cookieName);
+    }
+
+    public function removeToken(string $tokenId): ?string
+    {
+        return null;
+    }
+
+    public function isTokenValid(CsrfToken $token): bool
+    {
+        if (!$request = $this->requestStack->getCurrentRequest()) {
+            $this->logger?->debug('CSRF validation failed: No request found.');
+
+            return false;
+        }
+
+        if (false === $isValidOrigin = $this->isValidOrigin($request)) {
+            $this->logger?->debug('CSRF validation failed: origin info doesn\'t match.');
+
+            return false;
+        }
+
+        if (true === $isValidDoubleSubmit = $this->isValidDoubleSubmit($request, $token->getValue())) {
+            // Mark the request as validated using double-submit info
+            $request->attributes->set($this->cookieName, 'double-submit');
+            $this->logger?->debug('CSRF validation accepted using double-submit info.');
+
+            return true;
+        }
+
+        if (false === $isValidDoubleSubmit) {
+            $this->logger?->debug('CSRF validation failed: double-submit info doesn\'t match.');
+
+            return false;
+        }
+
+        // Opportunistically lookup at the session for a previous CSRF validation strategy
+        $session = $request->hasPreviousSession() ? $request->getSession() : null;
+        $usageIndexValue = $session instanceof Session ? $usageIndexReference = &$session->getUsageIndex() : 0;
+        $usageIndexReference = \PHP_INT_MIN;
+        $csrfProtection = $session?->get($this->cookieName);
+        $usageIndexReference = $usageIndexValue;
+
+        // If a previous request was validated using double-submit info, stick to it
+        if ('double-submit' === $csrfProtection && null === $isValidDoubleSubmit) {
+            $this->logger?->debug('CSRF validation failed: double-submit info was used in a previous request but is now missing.');
+
+            return false;
+        }
+
+        // If a previous request was validated using origin info, stick to it
+        if ('origin' === $csrfProtection && null === $isValidOrigin) {
+            $this->logger?->debug('CSRF validation failed: origin info was used in a previous request but is now missing.');
+
+            return false;
+        }
+
+        if (true === $isValidOrigin) {
+            // Mark the request as validated using origin info
+            $request->attributes->set($this->cookieName, 'origin');
+            $this->logger?->debug('CSRF validation accepted using origin info.');
+
+            return true;
+        }
+
+        $this->logger?->debug('CSRF validation failed: double-submit and origin info not found.');
+
+        return false;
+    }
+
+    public function clearCookie(Request $request, Response $response): void
```

Comment:
Is there any race condition risk that during a double-click form submission for example, the first response delete the cookie after the second one is generated but before the second request is sent ?

Reply:
Race condition indeed as the cookie name is unique.

The why that clear from PR description:
> The cookie is always cleared on the response to prevent any further use of the token.

OWASP [source](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#employing-hmac-csrf-tokens).
> A CSRF Token is not an access token.

The CSRF Token is not a one-time token nor expires.
It is not a weakness to reuse it.

The strength of this pattern is that on cross-site context, the double-submit is impossible with modern browsers.
Or I missed something.

Reply:
I don't think the race condition can really happen because of the way JS works (it's not multi-threaded and form submission don't happen in separate tabs at the same time).
Anyway, I updated the protocol so that this door is closed.
The name of the cookie is now the part that contains the token?